vshard/router/init.lua

```diff
@@ -576,11 +576,15 @@ local function router_call_impl(router, bucket_id, mode, prefer_replica,
     local call
     if mode == 'read' then
         if prefer_replica then
-            if balance then
+            if balance == "prefer_zone" then
```

Comment:
Please, change 'balance' to numeric IDs. String comparison in Lua not always works for constant time and might degrade to linear time. Like this:
```Diff
@@ -65,6 +65,10 @@ local fiber_yield = fiber.yield
 local fiber_cond_wait = util.fiber_cond_wait
 local gsc = util.generate_self_checker
 
+local BALANCE_NONE = 0
+local BALANCE_RR = 1
+local BALANCE_BEST_ZONE_RR = 2
+
 --
 -- on_connect() trigger for net.box
 --
@@ -578,7 +582,7 @@ local function replicaset_template_multicallro(prefer_replica, balance)
         local r
         local master = replicaset.master
 
-        if balance then
+        if balance ~= BALANCE_NONE then
             local prefered_zone = replicaset.priority_list[1].zone
             local i = #replicaset.priority_list
             while i > 0 do
@@ -588,7 +592,7 @@ local function replicaset_template_multicallro(prefer_replica, balance)
                     replica_check_backoff(r, now)
                 then
                     -- Pick a replica according prefered zone (highest priority replica zone) in round-robin manner
-                    if balance == "prefer_zone" and prefered_zone then
+                    if balance == BALANCE_BEST_ZONE_RR and prefered_zone then
                         local cbi = replicaset.balance_i
                         local nr = replicaset_balance_replica(replicaset)
 
@@ -1009,12 +1013,12 @@ local replicaset_mt = {
         wait_connected_all = replicaset_wait_connected_all,
         call = replicaset_master_call;
         callrw = replicaset_master_call;
-        callro = replicaset_template_multicallro(false, false);
-        callbro = replicaset_template_multicallro(false, true);
-        callbzro = replicaset_template_multicallro(false, "prefer_zone");
-        callre = replicaset_template_multicallro(true, false);
-        callbre = replicaset_template_multicallro(true, true);
-        callbzre = replicaset_template_multicallro(true, "prefer_zone");
+        callro = replicaset_template_multicallro(false, BALANCE_NONE);
+        callbro = replicaset_template_multicallro(false, BALANCE_RR);
+        callbzro = replicaset_template_multicallro(false, BALANCE_BEST_ZONE_RR);
+        callre = replicaset_template_multicallro(true, BALANCE_NONE);
+        callbre = replicaset_template_multicallro(true, BALANCE_RR);
+        callbzre = replicaset_template_multicallro(true, BALANCE_BEST_ZONE_RR);
         map_call = replicaset_map_call,
         update_master = replicaset_update_master,
     };
```

Reply:
Thank you for the comments. Will check it.